// userManagment/lib/userManagment.js
var db = require('dbHelper');
var login = function(userId,password,onSuccess,onFailure){
	db.query("select * from user where id='"+userId+"' and password='"+password+"'",function(error,results){
		if(error || results.length==0){
			onFailure();
		}else{
			onSuccess();
		}
	});
}

var createUserTable = function(){
	db.query("create table user(\n id varchar(50),\n password varchar(20),\n lastName varchar(20),\n firstName varchar(20),\n suffix varchar(10),\n description varchar(200),\nprimary key (id)\n)",function(error,results){});
}
var createUser = function(userId,password,onSuccess,onFailure){
	db.query("insert into user(id,password) values('"+userId+"','"+password+"')",function(error,results){
		if(error){
			onFailure();
		}else{
			onSuccess();
		}
	});
}

var setProfile = function(userId,lastName,firstName,suffix,description,onSuccess,onFailure){
	db.query("update user set lastName='"+lastName+"',firstName='"+firstName+"',suffix='"+suffix+"',description='"+description+"' where id='"+userId+"'",function(error,results){
		if(error){
			onFailure();
		}else{
			onSuccess();
		}
	});
}

var getProfile = function(userId,onSuccess,onFailure){
	db.query("select lastName,firstName,suffix,description from user where id='"+userId+"'",function(error,results){
		if(error){
			onFailure();
		}else{
			onSuccess(results[0]['lastName'],results[0]['firstName'],results[0]['suffix'],results[0]['description']);
		}
	});
}
var readProposal = function(roomNumber,onSuccess,onFailure){
	db.query("select description,options,vote from proposal where roomNumber='"+roomNumber+"'",function(error,results){
		if(error || results.length==0){
			onFailure();
		}else{
			onSuccess(results);
		}
	});
}
var insertProposal = function(roomNumber, description, options, onSuccess, onFailure){
	console.log("what?");
	var vote = "0";
	for(var i=1;i<options.split(":::").length;i++){
		vote = vote + ":::0";
	}
	db.query("insert into proposal(roomNumber,description,options,vote) values('"+roomNumber+"','"+description+"','"+options+"','"+vote+"')",function(error,results){
		if(error){
			onFailure();
		}else{
			onSuccess(results);
		}
	});
}
var increaseVote = function(roomNumber, description, voteResult, onSuccess, onFailure){
	db.query("select vote from proposal where roomNumber='"+roomNumber+"' and description='"+description+"'",function(error, results){
		if(error){
			onFailure();
		}else{
			var origin = results[0]["vote"].split(":::");
			var count = 0;
			var result = "";
			for(var i=0;i<origin.length;i++){
				if(voteResult[count]==i){
					result = result + (parseInt(origin[i])+1) + ":::";
					count++;
				}else{
					result = result + origin[i] + ":::";
				}
			}
			result = result.substring(0,result.length-3);
			db.query("update proposal set vote='"+result+"' where roomNumber='"+roomNumber+"' and description='"+description+"'",function(error, results){
				if(error){
					onFailure();
				}else{
					onSuccess();
				}
			});
		}
	});
}
exports.getProfile = getProfile;
exports.login = login;
exports.createUser = createUser;
exports.setProfile = setProfile;
exports.readProposal = readProposal;
exports.insertProposal = insertProposal;
exports.increaseVote = increaseVote;
